scripts/uploadToS3.js

```diff
@@ -0,0 +1,90 @@
+// @ts-check
+
+const chalk = require("chalk")
+const glob = require("glob")
+const mime = require("mime")
+const path = require("path")
+const s3 = require("s3")
+const { last } = require("lodash")
+
+const options = {
+  files: path.join(__dirname, "../**/public/**"),
+  root: "public",
+  key: process.env.S3_KEY,
+  secret: process.env.S3_SECRET,
+  bucket: "artsy-chris-test-bucket", // process.env.S3_BUCKET,  // artsy-force-$DEPLOY_ENV
+  cdnUrl: process.env.CDN_URL,
+}
+
+// See: https://www.npmjs.com/package/s3 for more options
+const client = s3.createClient({
+  s3Options: {
+    accessKeyId: options.key,
+    secretAccessKey: options.secret,
+  },
+})
+
+const files = glob.sync(options.files, { nodir: true }).filter(file => {
+  return !file.match("node_modules")
+})
+
+const generateHeaders = file => {
+  const contentType = mime.getType(
+    path.extname(
+      file
+        .replace(".br", "")
+        .replace(".gz", "")
+        .replace(".cgz", "")
+        .replace(".jgz", "")
+    )
+  )
+
+  let headers = {
+    CacheControl: "max-age=315360000, public",
+    ContentType: contentType,
+    ACL: "public-read",
+  }
+
+  if (file.match(/\.gz$/) || file.match(/\.cgz$/) || file.match(/\.jgz$/)) {
+    headers.ContentEncoding = "gzip"
+  }
+  if (file.match(/\.br$/)) {
+    headers.ContentEncoding = "br"
+  }
```

Comment:
hmm something is up here. When I run a prod build locally I get `br` and `gzip`, but when I point things at our S3 bucket no go 🤔

Reply:
I was wondering if we need to do some configuration to our Cloudfront CDN to serve assets as brotli... will investigate.

Reply:
Hrmm so Cloudfront will strip out all `Accept-Encoding` headers other than gzip :/  Lame.  Apparently there's a workaround using a Lambda hook but it will incur additional cost...
- https://medium.com/engineering-housing/how-we-served-brotli-compressed-files-through-amazon-cloudfront-b9414b70f372
- https://medium.com/@felice.geracitano/brotli-compression-delivered-from-aws-7be5b467c2e1
- https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/ServingCompressedFiles.html

Reply:
@izakp that last doc implies it can be configured, no?

> CloudFront can compress some types of files for you (see File Types that CloudFront Compresses), by using gzip. But if you want to compress other file types, or if you want to use a compression algorithm that isn’t gzip, such as brotli, you can compress the files on your own server, and then serve the files by using CloudFront.
>
> To use CloudFront to serve a file with a compression algorithm that isn’t gzip, set up CloudFront to cache based on the Accept-Encoding header. When you do this, CloudFront does not make any changes to the Accept-Encoding header and your origin can return an appropriate compressed file for the viewer.
> 
> When your origin returns a compressed file to CloudFront, include a Content-Encoding header to indicate to CloudFront that the file is already compressed. Then CloudFront simply returns the compressed file to the viewer.

Reply:
@joeyAghion that section of the docs relates to a Cloudfront distribution backed by a custom origin (i.e. Nginx) - if you look at the next section it describes the lambda workaround for a distribution backed by S3

Reply:
Bummer, that's unfortunate. We were getting a couple hundred k reductions on bundle sizes. The new artwork page was only 98k!

Reply:
It really would be good to figure out how to support this. I'm not really a fan of having all asset requests go through a lambda, but if it's just that one function I'd expect it stays plenty hot due to the traffic. Seems less overhead than managing a custom origin. We did talk about that before too though. I'm somewhat torn between the two approaches.

Reply:
@izakp - I'm sure this is a Cloudfront setup thing, but one thing i'm noticing is that when pointing local force at my test bucket assets _are not_ gzipped, even though they're being encoded as such and uploaded as such to S3, with the correct parameters. Running with local assets I get correct compression.